lib/Runtime/Library/JavascriptArray.cpp

```diff
@@ -6676,376 +6676,362 @@ using namespace Js;
-                {
-                    retVal = CALL_FUNCTION(threadContext, compFn, CallInfo(flags, 3), undefined, leftVar, rightVar);
-                }
-                END_SAFE_REENTRANT_CALL
-            }
-            else
-            {
-                BEGIN_SAFE_REENTRANT_CALL(threadContext)
-                {
-                    retVal = CALL_FUNCTION(scriptContext->GetThreadContext(), compFn, CallInfo(flags, 3), undefined, *(Var*)aRef, *(Var*)bRef);
-                }
-                END_SAFE_REENTRANT_CALL
-            }
+            retVal = CALL_FUNCTION(threadContext, compFn, CallInfo(flags, 3), undefined, leftVar, rightVar);
+        }
+        END_SAFE_REENTRANT_CALL
 
-            if (TaggedInt::Is(retVal))
-            {
-                return TaggedInt::ToInt32(retVal);
-            }
-            double dblResult;
-            if (JavascriptNumber::Is_NoTaggedIntCheck(retVal))
-            {
-                dblResult = JavascriptNumber::GetValue(retVal);
-            }
-            else
-            {
-                dblResult = JavascriptConversion::ToNumber_Full(retVal, scriptContext);
-            }
-            if (dblResult < 0)
-            {
-                return -1;
-            }
-            return (dblResult > 0) ? 1 : 0;
+        if (TaggedInt::Is(retVal))
+        {
+            return TaggedInt::ToInt32(retVal) < 0;
+        }
+        double dblResult;
+        if (JavascriptNumber::Is_NoTaggedIntCheck(retVal))
+        {
+            dblResult = JavascriptNumber::GetValue(retVal);
         }
         else
         {
-            JavascriptString* pStr1 = JavascriptConversion::ToString(*(Var*)aRef, requestContext);
-            JavascriptString* pStr2 = JavascriptConversion::ToString(*(Var*)bRef, requestContext);
-
-            return JavascriptString::strcmp(pStr1, pStr2);
+            dblResult = JavascriptConversion::ToNumber_Full(retVal, scriptContext);
         }
+        return dblResult < 0;
     }
 
-    static void hybridSort(__inout_ecount(length) Field(Var) *elements, uint32 length, CompareVarsInfo* compareInfo)
+    // Comparison method used in Array.prototype.sort with provided comparison function
+    // from same context as .sort call
+    bool compareVars(JavascriptArray::CompareVarsInfo* cvInfo, const void* aRef, const void* bRef)
     {
-        // The cost of memory moves starts to be more expensive than additional comparer calls (given a simple comparer)
-        // for arrays of more than 512 elements.
-        if (length > 512)
+        RecyclableObject* compFn=cvInfo->compFn;
+
+        AssertMsg(*(Var*)aRef, "No null expected in sort");
+        AssertMsg(*(Var*)bRef, "No null expected in sort");
+
+        ScriptContext* scriptContext = cvInfo->scriptContext;
+        ThreadContext* threadContext = scriptContext->GetThreadContext();
+        // The correct flag value is CallFlags_Value but we pass CallFlags_None in compat modes
```

Comment:
Not sure if I understand the comment - looks like we use the `flags` value in this case.

Reply:
The flags value we use (and this comment) were retained from the previous implementation - I didn't look into it, maybe a point for me to double check.

Reply:
Looked at it - that comment was in a bit of code I'd retained/copy+pasted from prior implementation. Looking at it AND other uses of the same comment (it appears in a few other places) it clearly seems to date back a long way and refer to a no longer existent "compat mode".

I'll add a commit that removes all of these comments.